### PR TITLE
FindOwners

### PR DESCRIPTION
**Pull request by cbart**
A partial implementation of `FindOwners` - which returns the owners defined in a CODEOWNERS file for given `path`.

I took a different approach here than in [the prototype](https://github.com/sourcegraph/sourcegraph/pull/45973) - which was heavily inspired by https://github.com/hmarr/codeowners/blob/main/match.go:

Instead of transpiling the glob pattern into a regular expression to match paths, we

- split the glob into parts by `/` separator,
- implement a simple matcher for each part,
- match file path by running a state machine.

**I truly do not know if this is a more readable approach as compared to building the regexp.**

On one hand building a massive regexp is tricky to get right, and might be hard to track the implementation. Plus it is one long method that is difficult to break down.

Here I propose a handful of small abstractions, that make the code way shorter. But given how much comment it required makes me doubt whether this is the right approach. **You tell me.**

This is missing one feature to keep the diff size down: it does not support yet pattern parts using `*` with text like, `*.md`.

## Test plan

Extensive table test cases - one with _matching_ samples, and the other one with _not matching_

### CODE REVIEW

File: internal/own/codeowners/proto/match_test.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+				// These do not match as the pattern matches anything within
+				// the sub-directory tree, but not the directory itself.
+				"/directory/path",
+				//"/prefix/directory/path",
+			},
+		},
+		{
+			pattern: "directory/*",
+			paths: []string{
+				//"/directory/nested/file",
```
Comment (indradhanush) [inline internal/own/codeowners/proto/match_test.go:96]: Intended to be commented out?

File: internal/own/codeowners/proto/match_test.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+				"/src/prefix_filename",
+			},
+		},
+		{
+			pattern: "directory/path/",
+			paths: []string{
+				"/directory/path_and_more/file",
+				"/prefix/directory/path_and_more/file",
+				// These do not match as the pattern matches anything within
+				// the sub-directory tree, but not the directory itself.
```
Comment (indradhanush) [inline internal/own/codeowners/proto/match_test.go:88]: Was the comment intended to be here or above the declaration of `cases`?

File: internal/own/codeowners/proto/match_test.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+				"/directory/path",
+				//"/prefix/directory/path",
```
Comment (indradhanush) [inline internal/own/codeowners/proto/match_test.go:90]: Commented out ?

File: internal/own/codeowners/proto/match_test.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+	codeownerspb "github.com/sourcegraph/sourcegraph/internal/own/codeowners/proto"
+)
+
+type testCase struct {
+	pattern string
+	paths   []string
+}
+
+func TestMatch(t *testing.T) {
```
Comment (indradhanush) [inline internal/own/codeowners/proto/match_test.go:15]: Maybe rename the function to reflect the function being tested? Like this or something similar?
```suggestion
func TestFileOwners_Match(t *testing.T) {
```

File: internal/own/codeowners/proto/match.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+// Visualized matching of `/src/java/test/UnitTest.java`:
+// / ** / src / java / test / ** / *Test.java   | Glob pattern
+// 0    1     2      3      4    5            6 | Bit vector index
+// *    *                                       | / (starting state)
+// *    *     *                                 | /src
+// *    *            *                          | /src/java
+// *    *                   *    *              | /src/java/test
+// *    *                   *    *            * | /src/java/test/UnitTest.java
+// The match is successful if after iterating throght the whole file path,
+// full pattern matches, that is there is a bit at the end of the glob.
```
Comment (indradhanush) [inline internal/own/codeowners/proto/match.go:77]: Love this! ❤️

File: internal/own/codeowners/proto/match.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+		case part == "":
+			return nil, errors.New("two consecutive forward slashes")
+		case part == "*":
+			p = append(p, anyMatch{})
+		case part == "**":
+			p = append(p, anySubPath{})
+		default:
+			p = append(p, exactMatch(part))
+		}
+	}
```
Comment (indradhanush) [inline internal/own/codeowners/proto/match.go:55]: You could rewrite this to remove the `part ==` from every case:

```suggestion
		switch part {
		case "":
			return nil, errors.New("two consecutive forward slashes")
		case "*":
			p = append(p, anyMatch{})
		case "**":
			p = append(p, anySubPath{})
		default:
			p = append(p, exactMatch(part))
		}
	}
```

**Review by indradhanush — COMMENTED** [review]
Left some suggestions. Not approved because another pair of eyes (or more) would be a good idea.

Not blocking because reading the code and comments I was able to follow the code for the most part but I also don't have a better suggestion  on how I'd do it myself. Sometimes I guess there's no way apart from the ugly way. The comments were amazing!

File: internal/own/codeowners/proto/match_test.go (hunk 4 above)
Comment (cbart) [reply internal/own/codeowners/proto/match_test.go]: Good idea. I'll do `TestFileOwnersMatch` without the underscore though. I recently got a suggestion (I think from @sashaostrikov) about the elegance of keeping the casing more consistent.

File: internal/own/codeowners/proto/match_test.go (hunk 2 above)
Comment (cbart) [reply internal/own/codeowners/proto/match_test.go:92]: No, it's actually intended for just the cases below. To lessen ambiguity, I added a similar comment to the paths above.

File: internal/own/codeowners/proto/match_test.go (hunk 3 above)
Comment (cbart) [reply internal/own/codeowners/proto/match_test.go]: OOps

File: internal/own/codeowners/proto/match_test.go (hunk 1 above)
Comment (cbart) [reply internal/own/codeowners/proto/match_test.go]: Ooops again

File: internal/own/codeowners/proto/match.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+func (p exactMatch) String() string         { return string(p) }
+func (p exactMatch) Match(part string) bool { return string(p) == part }
+
+// anyMatch is indicated by * in a glob pattern, and matches any single file
+// or directory on the path.
+type anyMatch struct{}
+
+func (p anyMatch) String() string         { return "*" }
+func (p anyMatch) Match(part string) bool { return true }
```
[the remark this answers is not in the document]
Comment (cbart) [reply internal/own/codeowners/proto/match.go]: Good good.

**Comment by cbart** [comment]
Thanks!

File: internal/own/codeowners/proto/match.go [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+			p = append(p, anySubPath{})
+		default:
+			p = append(p, exactMatch(part))
+		}
+	}
+	// Trailing `/` is equivalent with trailing `/**/*`.
+	// Such pattern matches any files within the directory sub-tree
+	// anchored at the director that the pattern describes.
+	if strings.HasSuffix(pattern, separator) {
+		p = append(p, anySubPath{}, anyMatch{})
```
[the remark this answers is not in the document]
Comment (sashaostrikov) [reply internal/own/codeowners/proto/match.go:90]: I've tried to remove it and ran tests -- they passed 🤔

**Review by sashaostrikov — APPROVED** [review]
Very nice! Left some comments, which can be basically fixed without an additional round of reviews, so approving right away :)

In any case, if you want another round -- just tag/DM me!

File: internal/own/codeowners/proto/match.go [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+	}
+	return p, nil
+}
+
+// match iterates over `filePath` separated by `/`. It uses a bit vector
+// to track which prefixes of glob pattern match the file path prefix so far.
+// Bit vector indices correspond to separators between pattern parts.
+// Visualized matching of `/src/java/test/UnitTest.java`:
+// / ** / src / java / test / ** / *Test.java   | Glob pattern
```
[the remark this answers is not in the document]
Comment (cbart) [reply internal/own/codeowners/proto/match.go:104]: 👍

File: internal/own/codeowners/proto/match.go [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+// globPattern implements a pattern for matching file paths,
+// which can use directory/file names, * and ** wildcards,
+// and may or may not be anchored to the root directory.
+type globPattern []patternPart
+
+// compile translates a text representation of a glob pattern
+// to an executable one that can `match` file paths.
+func compile(pattern string) (globPattern, error) {
+	var p []patternPart
```
[the remark this answers is not in the document]
Comment (cbart) [reply internal/own/codeowners/proto/match.go]: 👍

File: internal/own/codeowners/proto/match.go [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// FindOwners returns the Owners associated with given path as per this CODEOWNERS file.
+// Rules are evaluated in order: Returned owners come from the rule which pattern matches
+// given path, that is the furthest down the file.
+func (f *File) FindOwners(path string) []*Owner {
+	var owners []*Owner
+	for _, r := range f.GetRule() {
+		m, err := compile(r.GetPattern())
```
[the remark this answers is not in the document]
Comment (cbart) [reply internal/own/codeowners/proto/match.go]: 👍

File: internal/own/codeowners/proto/match.go [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// FindOwners returns the Owners associated with given path as per this CODEOWNERS file.
+// Rules are evaluated in order: Returned owners come from the rule which pattern matches
+// given path, that is the furthest down the file.
+func (f *File) FindOwners(path string) []*Owner {
+	var owners []*Owner
+	for _, r := range f.GetRule() {
```
[the remark this answers is not in the document]
Comment (cbart) [reply internal/own/codeowners/proto/match.go]: 👍

File: internal/own/codeowners/proto/match.go [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+	"math/big"
+	"strings"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// FindOwners returns the Owners associated with given path as per this CODEOWNERS file.
+// Rules are evaluated in order: Returned owners come from the rule which pattern matches
+// given path, that is the furthest down the file.
+func (f *File) FindOwners(path string) []*Owner {
```
[the remark this answers is not in the document]
Comment (cbart) [reply internal/own/codeowners/proto/match.go]: 👍

**Review by cbart — COMMENTED** [review]
Thanks for the very thorough review! All comments addressed. WIll be 🛬 .

File: internal/own/codeowners/proto/match.go [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+// / ** / src / java / test / ** / *Test.java   | Glob pattern
+// 0    1     2      3      4    5            6 | Bit vector index
+// *    *                                       | / (starting state)
+// *    *     *                                 | /src
+// *    *            *                          | /src/java
+// *    *                   *    *              | /src/java/test
+// *    *                   *    *            * | /src/java/test/UnitTest.java
+// The match is successful if after iterating throght the whole file path,
+// full pattern matches, that is there is a bit at the end of the glob.
+func (p globPattern) match(filePath string) bool {
```
[the remark this answers is not in the document]
Comment (cbart) [reply internal/own/codeowners/proto/match.go]: So the constructor `compile` goes first, then type definition, then methods - as opposed to type definition, then constructor, then methods. I'm game.

File: internal/own/codeowners/proto/match.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (cbart) [reply internal/own/codeowners/proto/match.go]: Very much appreciate naming police. I do go a bit crazy with single-letter-naming, and I'm glad someone keeps an eye on that.

**Pull request merged**